magicctapipe/scripts/lst1_magic/lst_m1_m2_cloud_correction.py

```diff
@@ -0,0 +1,417 @@
+
+        tel_frame = TelescopeFrame(telescope_pointing=telescope_pointing)
+        tel = pointing_altaz.transform_to(tel_frame)
+
+        src_x = tel.fov_lat.to(u.deg).value
+        src_y = tel.fov_lon.to(u.deg).value
+
+        # Transform to Engineering camera
+        src_x, src_y = -src_y * u.deg, -src_x * u.deg
+        cog_x, cog_y = -cog_y, -cog_x
+
+        pix_x_tel = (camgeom[telid].pix_x * m2deg).to(u.deg)
+        pix_y_tel = (camgeom[telid].pix_y * m2deg).to(u.deg)
+
+        distance = np.abs(
+            (pix_y_tel - src_y) * np.cos(psi) + (pix_x_tel - src_x) * np.sin(psi)
+        )
+
+        d2_cog_src = (cog_x - src_x) ** 2 + (cog_y - src_y) ** 2
+        d2_cog_pix = (cog_x - pix_x_tel) ** 2 + (cog_y - pix_y_tel) ** 2
+        d2_src_pix = (src_x - pix_x_tel) ** 2 + (src_y - pix_y_tel) ** 2
+
+        distance[d2_cog_pix > d2_cog_src + d2_src_pix] = 0
+        dist_corr_layer = model2(impact, Hcl, zenith) * u.deg
+
+        ilayer = np.digitize(distance, dist_corr_layer)
+        trans_pixels = transl[ilayer]
+
+        inds_img = np.where(
+            (dl1_images["event_id"] == event_id)
+            & (dl1_images["tel_id"] == telid)
+            & (dl1_images["obs_id"] == obs_id)
+        )[0]
+
+        if len(inds_img) > 0:
+            for index_img in inds_img:
+                image = dl1_images["image"][index_img]
+                cleanmask = dl1_images["image_mask"][index_img]
+                peak_time = dl1_images["peak_time"][index_img]
+                image /= trans_pixels
+                corr_image = image.copy()
+                corr_image[~cleanmask] = 0
+
+                hillas_params = hillas_parameters(camgeom[telid], corr_image)
+                timing_params = timing_parameters(
+                    camgeom[telid], corr_image, peak_time, hillas_params
+                )
+                leakage_params = leakage_parameters(
+                    camgeom[telid], corr_image, signal_pixels
+                )
+                conc_params = concentration_parameters(
+                    camgeom[telid], corr_image, hillas_params
+                )
+
+                source_pos_in_camera = sky_to_camera(
+                    alt_rad * u.rad,
+                    az_rad * u.rad,
+                    focal_eq[telid],
+                    pointing_alt * u.rad,
+                    pointing_az * u.rad,
+                )
+
+                disp_parameters = disp.disp(
+                    x, y, source_pos_in_camera.x, source_pos_in_camera.y, psi
+                )
+
+                disp_dx = disp_parameters[0].value * u.m
+                disp_dy = disp_parameters[1].value * u.m
+                disp_norm = disp_parameters[telid].value * u.m
+                disp_angle = disp_parameters[3].value * u.rad
+                disp_sign = disp_parameters[4]
+                src_x = source_pos_in_camera.x.value * u.m
+                src_y = source_pos_in_camera.y.value * u.m
+
+                data.append(
```

Comment:
as we were discussing specifing all the parameters by hand makes the code unnecessarily complicated, much harder to maintain and prone to errors (there are some below)
you should save the whole containers, here you have an example how to do it:
https://github.com/cta-observatory/magic-cta-pipe/blob/352c56fdf11237322c41f055f0c976f4b34ef8fb/magicctapipe/scripts/lst1_magic/magic_calib_to_dl1.py#L394-L397

you will just need to stack them up from all the telescopes

Reply:
Now I'm saving the parameters with pandas:

save_pandas_data_in_table(
    df_all, output_file, group_name="/events", table_name="parameters"
)